docs/design/federated-services.md

```diff
@@ -0,0 +1,318 @@
+            name: my-service
+            ports:
+            - containerPort: 2379
+              protocol: TCP
+            - containerPort: 2380
+              protocol: TCP
+
+Ubernetes in turn creates one equivalent replication controller (identical config to the above, except for the replica count) in each of the underlying Kubernetes clusters, each of which results in something like this:
+
+    $ ./kubectl get -o yaml rc my-service --context="cluster-1"
+    kind: ReplicationController
+    metadata:
+      creationTimestamp: 2015-12-02T23:00:47Z
+      labels:
+        run: my-service
+      name: my-service
+      namespace: my-namespace
+      selfLink: /api/v1/namespaces/my-namespace/replicationcontrollers/my-service
+      uid: 86542109-9948-11e5-a38c-42010af00002
+    spec:
+      replicas: 2
+      selector:
+        run: my-service
+      template:
+        metadata:
+          labels:
+            run: my-service
+        spec:
+          containers:
+            image: gcr.io/google_samples/my-service:v1
+            name: my-service
+            ports:
+            - containerPort: 2379
+              protocol: TCP
+            - containerPort: 2380
+              protocol: TCP
+            resources: {}
+          dnsPolicy: ClusterFirst
+          restartPolicy: Always
+    status:
+      replicas: 2
+
+The exact number of replicas created in each underlying cluster will of course depend on what scheduling policy is in force.  In the above example, the scheduler created an equal number of replicas (2) in each of the three underlying clusters, to make up the total of 6 replicas required. 
+
+### Implementation Details
+
+The implementation approach and architecture is very similar to Kubernetes, so if you're familiar with how Kubernetes works, none of what follows will be surprising.  One additional design driver not present in Kubernetes is that Ubernetes aims to be resilient to individual cluster and availability zone failures.  So the control plane spans multiple clusters.  More specifically:
+
++  Ubernetes runs it's own distinct set of API servers (typically one or more per underlying Kubernetes cluster).  These are completely distinct from the Kubernetes API servers for each of the underlying clusters.
++  Ubernetes runs it's own distinct quorum-based metadata store (etcd, by default).  Approximately 1 quorum member runs in each underlying cluster ("approximately" because we aim for an odd number of quorum members, and typically don't want more than 5 quorum members, even if we have a larger number of federated clusters, so 2 clusters->3 quorum members, 3->3, 4->3, 5->5, 6->5, 7->5 etc).  
+"Ubelets" in Ubernetes work much like "kubelet" in Kubernetes.  They watch against the Ubernetes API server/etcd state, and apply changes to the underlying kubernetes clusters accordingly.  They also have the anti-entropy mechanism for reconciling ubernetes "desired desired" state against kubernetes "actual desired" state.
```

Comment:
Correct me if I'm on the wrong track here, but my understanding is that kubelet watches the desired state (in etcd, via API server) of all pods scheduled onto it, and manipulates the state on the node to match that.  This is the sense in which I meant that Ubeletes are much like Kubeletes.  The perhaps confusing part is the "desired desired" state bit (no that's not a typo :-)  ). What's important to note in that regard is that Ubelets manipulate desired state in the underlying clusters (e.g. the desired number of replicas of in cluster X), rather than the actual state there. This desired state is distinct from the desired federation state as a whole (e.g.